### ISSUE TITLE
Representing "structs"

### ISSUE THREAD

**Issue by skyne98** [CraneStation/cranelift#638]
Hello there, I have a slightly trivial question.

How does one represent a structure in Cranelift IR? I know that structures are more-or-less a virtual concept and can be introduced "virtually", by representing parts of them as different objects in memory. However, what if one wants to inter-operate with C library APIs that require passing a structure? As far as I understand, some strict memory layout standard should be followed with correct paddings and organisation. 

Is there any easy way of doing this? Thanks!

**Comment by sunfishcode** [comment]
You'll need to calculate the size and alignment of the struct, to know how much memory to allocate, and the offsets of the fields, to know how to access the fields. Cranelift doesn't currently provide an automatic way to do this, though it is something we could consider. In many simple cases, it's straightforward -- you just append one field at a time, inserting padding at each step as needed to ensure the alignment of the field.

Depending on your needs, it may be possible to get the Rust compiler to do the work for you. If you can declare the structs you need as extern "C" structs in Rust, you can then use `mem::size_of`, `mem::align_of`, and [the memoffset crate](https://crates.io/crates/memoffset) to calculate field offsets.

**Comment by skyne98** [comment]
I think that supporting both structures and arrays is essential to avoid projects making their own implementations all the time -- this should help avoid a lot of confusion, increase accessibility and introduce only one single point of failure for bugs to appear.

Also, introducing them will mean that each improvement will benefit all the other projects (I think we all know the benefits of reusing code).

As for right now, I totally understand that WebAssembly compilers (which is the main use of Cranelift at the moment) do not require it to have such features, but, again, introducing them will help other projects, that have to think about such things.

**Comment by blanham** [comment]
@skyne98 I totally agree with you. Having that functionality built-in would increase the adoption of Cranelift substantially. From looking at various toy compilers, most of them forgo structures because of the added complexity (and there aren't many examples for someone new to compilers, things like LLVM's Kaleidoscope tutorial completely ignore handling data structures).

Though I also buy that this could be in a separate crate. What I'd really love is an equivalent to https://github.com/chriswailes/RLTK but in Rust

**Comment by sunfishcode** [comment]
RLTK looks like a cool package.

There are a variety of parser crates for Rust, such as [lalrpop](https://crates.io/crates/lalrpop) and others. The [simplejit-demo](https://github.com/CraneStation/simplejit-demo) is using [peg](https://crates.io/crates/peg). It'd be cool to put some of these pieces together to start building something like RLTK. I unfortunately don't have time to do this right now, but I'd be happy to help anyone else who wanted to!

**Comment by skyne98** [comment]
@sunfishcode, I have found some good articles regarding basic struct representations in memory. At the moment I am experimenting with implementing them in a toy language compiler.

If it does work, I think it might be a good idea to try to integrate it into Cranelift with your overseeing and mentorship.

**Comment by sunfishcode** [comment]
@skyne98 That sounds great! I'd be happy to help mentor such a project.
